### ISSUE TITLE
Use the gamma distribution for ancestral sequence reconstruction

### ISSUE THREAD

**Issue by BenitezG** [veg/hyphy#551]
Hello,
I'm doing a study to compare different evolutionary models with different parameters (like gamma distribution or observed frequencies) in the ancestral sequence reconstruction (ASR) and I'm having a problem with the gamma distribution. The idea of this study is to simulate an evolutionary history via seq-gen software. I give the seq-gen a series of trees with fixed topology and a particular evolutionary model and I obtain the sequences of the nodes. After that I use the sequences from the leaves of the trees and make the ASR with the HYPHY. For this purpose I made this script:
```
DataSet data = ReadDataFile ("seqfile");

ACCEPT_ROOTED_TREES = 1;
AUTOMATICALLY_CONVERT_BRANCH_LENGTHS = 1;
LIKELIHOOD_FUNCTION_OUTPUT=0;

UseModel (USE_NO_MODEL);

/*create a partition in base of the alignment*/

        DataSetFilter clase = CreateFilter (data,1);

/*Define model global*/

/*JONES*/
       global alpha = 1;
       alpha:>0.01;alpha:<100;
       category c = (4, EQUAL, MEAN, GammaDist(_x_,alpha,alpha),CGammaDist(_x_,alpha,alpha),0,1e25,CGammaDist(_x_,alpha+1,alpha));
#include "Jones.mdl";

Tree arbol = (uno:5.003,((dos:5.001,tres:5.001):5.001,(cuatro:5.001,cinco:5.001):5.001):5.001);

/*Define the likelihood function*/

        LikelihoodFunction lk = (clase, arbol, SITE_LIKELIHOOD);


        fprintf (mljtt-jttG, " Jones Log(L) = ", lk);
        fprintf (mljtt-jttG, " uno = ", BranchLength(arbol,"uno"),"\n");
        fprintf (outs, "Jones unconstrained  Log(L) = ", lk,"\n");

/*ASR*/

        DataSet ancestors = ReconstructAncestors (lk,MARGINAL);

        DataSetFilter ancestral_filter = CreateFilter (ancestors, 1);
        fprintf (outs, "Inferred\n", ancestors, "\n");

        GetDataInfo (node0, ancestral_filter, 0);
        fprintf (outs, "node0 ", node0, "\n");
        GetDataInfo (node2, ancestral_filter, 1);
        fprintf (outs, "node2 ", node2, "\n");
        GetDataInfo (node3, ancestral_filter, 2);
        fprintf (outs, "node3 ", node3, "\n");
        GetDataInfo (node6, ancestral_filter, 3);
        fprintf (outs, "node6 ", node6, "\n");

```

The topology of the tree vary from branch length = 0.001 to 1.400
To make the comparison I calculate the identity percentage of the sequences (the node obtained from seq-gen vs the node obtained from HYPHY).
The problem is that the results of these analysis are the same that the obtained when I don't use the gamma distribution, i.e. the identity percentages for each branch length are the same.
I expected the results to vary among the ASR with and without the gamma distribution, so I think that I'm doing something wrong.
I attach the Jones.mdl file, that is the matrix of the model, which I made based on the Jones.mdl file that comes with the HYPHY.
[Jones.mdl.txt](https://github.com/veg/hyphy/files/1024165/Jones.mdl.txt)

**Comment by spond** [comment]
Dear @BenitezG, 

Try this 

```
DataSet data = ReadDataFile ("seqfile");

ACCEPT_ROOTED_TREES = 1;
LIKELIHOOD_FUNCTION_OUTPUT=0;
VERBOSITY_LEVEL = 0;

UseModel (USE_NO_MODEL);

/*create a partition in base of the alignment*/

        DataSetFilter clase = CreateFilter (data,1);

/*Define model global*/

/*JONES*/
       global alpha = 1;
       alpha:>0.01;alpha:<100;
       category c = (4, EQUAL, MEAN, GammaDist(_x_,alpha,alpha),CGammaDist(_x_,alpha,alpha),0,1e25,CGammaDist(_x_,alpha+1,alpha));
       
categoriesUsed = 1;      
       
#include "Jones.mdl";

Tree arbol = (uno:5.003,((dos:5.001,tres:5.001):5.001,(cuatro:5.001,cinco:5.001):5.001):5.001);

/*Define the likelihood function*/

        LikelihoodFunction lk = (clase, arbol, SITE_LIKELIHOOD);

        Optimize (res, lk);

        fprintf (mljtt-jttG, " Jones Log(L) = ", lk);
        fprintf (mljtt-jttG, " uno = ", BranchLength(arbol,"uno"),"\n");
        fprintf (stdout, "Jones unconstrained  Log(L) = ", lk,"\n");

/*ASR*/

        DataSet ancestors = ReconstructAncestors (lk,MARGINAL);

        DataSetFilter ancestral_filter = CreateFilter (ancestors, 1);
        fprintf (outs, "Inferred\n", ancestors, "\n");

        GetDataInfo (node0, ancestral_filter, 0);
        fprintf (stdout, "node0 ", node0, "\n");
        GetDataInfo (node2, ancestral_filter, 1);
        fprintf (stdout, "node2 ", node2, "\n");
        GetDataInfo (node3, ancestral_filter, 2);
        fprintf (stdout, "node3 ", node3, "\n");
        GetDataInfo (node6, ancestral_filter, 3);
        fprintf (stdout, "node6 ", node6, "\n");
```

Best,
Sergei

**Comment by BenitezG** [comment]
Thank you for answering so quickly. 
This works, but it modifies the tree topology and I need to have a fixed tree topology. This occurs because the following command is removed from the script:
`AUTOMATICALLY_CONVERT_BRANCH_LENGTHS = 1;`
And the optimize command is added.
`Optimize (res, lk);`
The problem is that when I remove the optimize command and add the "AUTOMATICALLY_CONVERT_BRANCH_LENGHTS" the output generated by HyPhy with the script gives me the reconstructed nodes, but they're all alanines and the likelihood isn't calculated.
```
WAG unconstrained  Log(L) =             inf
Inferred
4 species: {Node0,Node2,Node3,Node6};
Total Sites: 250;
Distinct Sites: 1
node0 AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
node2 AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
node3 AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
node6 AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
```
Also, the tree topology is modified even though the "AUTOMATICALLY_CONVERT_BRANCH_LENGHTS" command is added!!
This topology is very strange. The value of branch lenghts is 3.026e+26!! 
I don't know why this occurs.
Is there a way by which the tree topology stays fixed and the likelihood is correctly calculated? 
Thank you.

**Comment by spond** [comment]
Dear @BenitezG,

You mean you want the branch lengths to be fixed? 

`AUTOMATICALLY_CONVERT_BRANCH_LENGTHS` seems to have a bug which prevents it from correctly setting the branch length parameter in the presence of rate variation. 

Try this

```
DataSet data = ReadDataFile ("seqfile");

ACCEPT_ROOTED_TREES = 1;
LIKELIHOOD_FUNCTION_OUTPUT=0;

/*create a partition in base of the alignment*/

        DataSetFilter clase = CreateFilter (data,1);

/*Define model global*/

/*JONES*/
       global alpha = 1;
       alpha:>0.01;alpha:<100;
       category c = (4, EQUAL, MEAN, GammaDist(_x_,alpha,alpha),CGammaDist(_x_,alpha,alpha),0,1e25,CGammaDist(_x_,alpha+1,alpha));
       
categoriesUsed = 1;      

tree_string = "(uno:5.003,((dos:5.001,tres:5.001):5.001,(cuatro:5.001,cinco:5.001):5.001):5.001)";

UseModel (USE_NO_MODEL);
Tree ref_tree = tree_string;

#include "Jones.mdl";

Tree arbol = tree_string;
GetString (branch_length_expression, JonesModel, -1);
branch_count = BranchCount (arbol) + TipCount (arbol);

c = 1;
branch_names = BranchName (ref_tree, -1);

for (branch = 0; branch < branch_count; branch += 1) {
    user_value  = BranchLength (ref_tree, branch);
    ExecuteCommands ("FindRoot (parameter_value, `branch_length_expression` - `user_value`, t, 0, 10000)");
    *("arbol.`branch_names[branch]`.t") := parameter_value__;
}

/*Define the likelihood function*/

        LikelihoodFunction lk = (clase, arbol, SITE_LIKELIHOOD);

        Optimize (res, lk);

        fprintf (mljtt-jttG, " Jones Log(L) = ", lk);
        fprintf (mljtt-jttG, " uno = ", BranchLength(arbol,"uno"),"\n");
        fprintf (stdout, "Jones unconstrained  Log(L) = ", lk,"\n");

/*ASR*/

        DataSet ancestors = ReconstructAncestors (lk,MARGINAL);

        DataSetFilter ancestral_filter = CreateFilter (ancestors, 1);
        fprintf (outs, "Inferred\n", ancestors, "\n");

        GetDataInfo (node0, ancestral_filter, 0);
        fprintf (stdout, "node0 ", node0, "\n");
        GetDataInfo (node2, ancestral_filter, 1);
        fprintf (stdout, "node2 ", node2, "\n");
        GetDataInfo (node3, ancestral_filter, 2);
        fprintf (stdout, "node3 ", node3, "\n");
        GetDataInfo (node6, ancestral_filter, 3);
        fprintf (stdout, "node6 ", node6, "\n");
```

Best,
Sergei

**Comment by BenitezG** [comment]
Dear @spond 
I'm sorry to keep bothering you so much, but when I run this script this error appears:
```
Error:

Master node received an error:Insufficient number of arguments for a call to Operation -

Function call stack
1 : Find the root of 0.100643171103352*c*t- in t between 0 and 10000 storing the result in parameter_value
-------
2 : ExecuteCommands in string "FindRoot (parameter_value, `branch_length_expression` - `user_value`, t, 0, 10000)" using basepath /home/guillermo/Escritorio/JTT-JTT+Ga/.
-------
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
```
I printed branch_length_expression and user_value's values
branch_length_expression = 0.100643171103352*c*t
user_value= 0.003
It's like the function FindRoot didn't read the user_value, because in the error message it doesn't appears.

**Comment by spond** [comment]
Dear @BenitezG,

Please try running this using v2.3-dev branch. 

```
$git clone https://github.com/veg/hyphy.git hyphy-2.3
$cd hyphy-2.3
$git checkout v2.3-dev
$cmake ./
$make MP
$./HYPHYMP LIBPATH=`pwd`/res 
```

The errors you are seeing appear to be due to the fact that the code that I wrote uses some features which are not available in v2.2* builds.

Best,
Sergei

**Comment by BenitezG** [comment]
It worked perfectly.
Thank you very much for all your help!
